### PR TITLE
feat(tapac): support top-level module having istream/ostream connections. Also, fixing bug when a design has no buffers or no fifos.

### PR DESCRIPTION
**Pull request by KennySLiu**
This was inspired by Kenny trying to add QSFP-support to a kernel, which requires interfacing with an RTL kernel.

Note that the top-level having istream/ostream connections does NOT Work with autobridge. So users will have to disable floorplanning, if they use top-level stream connections, until we come up with a fix .

### CODE REVIEW

**Comment by KennySLiu** [comment]
The relevant change (in tapa/backend/python/tapa/verilog/xilinx/\_\_init\_\_.py) relies on the [haoda repository](https://github.com/Blaok/haoda/blob/master/haoda/backend/xilinx.py#L165), which defines a bunch of tcl commands necessary to package the XO file. 

The haoda repo supports defining interfaces, but it's unused by TAPA. I simply added usage of the interface support.

File: tapa/backend/python/tapa/task.py
Comment (thecurryspice) [inline tapa/backend/python/tapa/task.py:1]: @KennySLiu, can you please explain why this change is needed?
If the original code needed to return an empty dictionary `{}`, then we should have expected the program to throw a `KeyError`.

Since `kwargs` already contains a dictionary of child fifo and buffer names, I don't understand why this change is required.

File: tapa/backend/python/tapa/task_graph.py
Comment (thecurryspice) [inline tapa/backend/python/tapa/task_graph.py:1]: Would this not conflict with any other tasks that have streams incorrectly defined?
(I am happy to accept this change with the asserted logger.warning if this is what enables the QSFP communication)

File: tapa/backend/python/tapa/task.py
Comment (KennySLiu) [reply tapa/backend/python/tapa/task.py:1]: ~~Oh, good point. This change is not needed, I added it because tapac was crashing at this point and I didn't know why. I wanted to see what the exception that was being thrown was.~~

File: tapa/backend/python/tapa/task_graph.py
Comment (KennySLiu) [reply tapa/backend/python/tapa/task_graph.py:1]: I just realized I misunderstood how the GitHub UI places comments in reference to changes. I thought the comment was *below* the relevant change, but it's actually above the relevant change. Below is my updated response.

File: tapa/backend/python/tapa/task.py
Comment (KennySLiu) [reply tapa/backend/python/tapa/task.py:1]: (((I just realized I misunderstood how the GitHub UI places comments in reference to changes. I thought the comment was *below* the relevant change, but it's actually above the relevant change. Therefore the above comment addresses a different code change. Here's my actual response.)))

From what I remember, the original code didn't handle programs that had no tapa buffer instantiations (i.e. BitBlender). And, I assumed that it also wouldn't handle programs that had no tapa stream instantiations. 

So, I found that a design that defines no buffers would pass `kwargs` without the key "buffers". As a result, the `.pop("buffers")` gives a KeyError. If this is wrong, then I don't know what went wrong with my design. I'll need to look more into it.

But assuming that assumption is correct, here's a python example of what happens without my fix, for a design that defines no buffers.

```
import collections

empty_dict = {}
buffs = collections.OrderedDict(
            sorted( 
                    (item for item in empty_dict.pop("buffers").items()) ,
                    key = lambda x: x[0]
            ) 
        )
````

If this KeyError is required for some other error checking, then I suppose PASTA would need to require users to define at least one dummy buffer (and FIFO)?

File: tapa/backend/python/tapa/task_graph.py
Comment (KennySLiu) [reply tapa/backend/python/tapa/task_graph.py:1]: You're right, this probably would conflict with incorrectly-defined stream declarations. I suppose there would 
 be another, better way to do this, but I don't know what it would be.

Also, note that this change is required for AutoBridge to not crash. If floorplanning is disabled, this change is NOT required.

**Comment by KennySLiu** [comment]
Can we merge these changes?

**Comment by thecurryspice** [comment]
This looks great! Thanks!

**Review by thecurryspice — APPROVED** [review]
